### PR TITLE
Merge request/response HTTP headers in root and delegate VS

### PR DESCRIPTION
**Pull request by varks**
Request/Response HTTP headers in root VS and delegate VS are not currently merged. 

This PR adds support to merge request/response headers in root and delegate VS based on merge semantics defined in the virtual service chaining [doc](https://docs.google.com/document/d/1TzXjvbEqeRKHQ7FonAY_yS1zHa8rVH2DIq4EM5BuDQ4/edit#heading=h.udcizzgca16m)

I am not sure if this needs to be added to release notes. Please advise.

### CODE REVIEW

File: pilot/pkg/model/virtualservice.go [hunk 1]
```diff
@@ -413,6 +412,109 @@ func mergeHTTPMatchRequest(root, delegate *networking.HTTPMatchRequest) *network
...
+	}
+	// request header operations
+	if root.Request == nil {
+		out.Request = delegate.Request
+	}
+	if delegate.Request == nil {
+		out.Request = root.Request
+	}
+	if root.Request != nil && delegate.Request != nil {
+		// Set - merge the map if not intersected, otherwise delegators value will
```
Comment (ramaraochavali) [inline pilot/pkg/model/virtualservice.go:488]: nit: delegators - delegates

**Review by ramaraochavali — COMMENTED** [review]
LGTM. Minor comment. @hzxuzhonghu  for header merging for delegates that we discussed offline - Can you PTAL?

**Review by hzxuzhonghu — COMMENTED** [review]
LGTM, just imagine if a root vs with headers remove: ["header1"], and the delegate vs with hraders add["header1": "value1"].  I think for this case the merge may lead to "header1" removed.

**Comment by ramaraochavali** [comment]
>  LGTM, just imagine if a root vs with headers remove: ["header1"], and the delegate vs with hraders add["header1": "value1"]. I think for this case the merge may lead to "header1" removed.

Good point. I did not think about this. Should we process removes first?

File: pilot/pkg/model/virtualservice.go (hunk 1 above)
Comment (varks) [reply pilot/pkg/model/virtualservice.go]: fixed.

**Comment by ramaraochavali** [comment]
> I don't think we should merge because of this

Is this the reason why it was not implemented originally - I could not remember the history

I think there is actually valid use case for merge when delegates are used between Gateway and Mesh especially around new headers - We need to keep both. One way to solve that is to always add headers in Gateway Root VS but that is not possible in all cases. Other options are using custom envoy filters like Lua. May be we can support this with that caveat that remove/add of same header can be undeterministic?

**Comment by varks** [comment]
@howardjohn We would like to get this change in as we have a use case which @ramaraochavali shared above. When you get a chance, could you take a look again?. Thanks.

**Pull request closed without merging**